### ISSUE TITLE
About mask normalization

### ISSUE THREAD

**Issue by dkoguciuk** [JirongZhang/DeepHomography#19]
Hi @JirongZhang,

I am absolutely happy you manage to open the code! I would probably ask you about the code in the following days, and my first question is about mask normalization.

This is the function:
```
def normMask(mask, strenth = 0.5):
    """
    :return: to attention more region

    """
    batch_size, c_m, c_h, c_w = mask.size()
    max_value = mask.reshape(batch_size, -1).max(1)[0]
    max_value = max_value.reshape(batch_size, 1, 1, 1)
    mask = mask/(max_value*strenth)
    mask = torch.clamp(mask, 0, 1)

    return mask
```

And this is the invocation:
```
mask_I1 = normMask(mask_I1)
mask_I2 = normMask(mask_I2)

...

pred_Mask = normMask(pred_Mask)
```

So, as far as I understand, for each mask prediction, you normalize it with the half of the maximum value. This will result in normalizing the mask in the range (0, 2), but in the last line, you clamp the mask values to be the range of (0, 1). And my question is: **why?** I have some naive intuition behind it, but I am not fully convinced by this intuition. And, why didn't you mention it in the paper? I don't see it as a simple implementation trick.

**Comment by JirongZhang** [comment]
Thank you. 
When the mask is not normalized, a limited number of pixels are highlighted, but the images can still be aligned. Intuitively, I want the mask to focus on more areas, so I normalized the mask. At present, we lack the discussion of this function and the "strenth" parameter.

**Comment by JirongZhang** [comment]
Thank you.
When the mask is not normalized, a limited number of pixels are highlighted, but the images can still be aligned. Intuitively, I want the mask to focus on more areas, so I normalized the mask. At present, we lack the discussion of this function and the "strenth" parameter. @dkoguciuk

**Comment by dkoguciuk** [comment]
Thanks @JirongZhang for clarification. I've also seen that features are extracted based on the patches:

```
        patch_1 = self.ShareFeature(input_tesnors[:, :1, ...])
        patch_2 = self.ShareFeature(input_tesnors[:, 1:, ...])
```

but masks are estimated for the whole input image, and then are cropped to match the size of input tensors/features:

```
        mask_I1_full = self.genMask(org_imges[:, :1, ...])
        mask_I2_full = self.genMask(org_imges[:, 1:, ...])

        mask_I1 = getPatchFromFullimg(patch_size_h, patch_size_w, patch_indices, batch_indices_tensor, mask_I1_full)
        mask_I2 = getPatchFromFullimg(patch_size_h, patch_size_w, patch_indices, batch_indices_tensor, mask_I2_full)
```

What is the intuition behind it? Have you tested this against estimating mask for input tensors only, and not the whole original images? Now, the network can use the area outside `patch_2` to match `patch_1` features. Is it desirable? Or should we use only the visible part of pre-transformed `patch_2`?

Did I express myself clearly, or should I prepare some example figures illustrating my doubt here?

**Comment by JirongZhang** [comment]
For the feature, we first extract the features of the patch. After getting h, we warp the original image and crop it to get the new patch, **and then extract the feature from the new patch**.
For the mask, we first make mask prediction on the original image, and then cut the prediction results to fit the patch size. After H is obtained,  we warp the whole mask and then cut it to fit the tensor size.
There are some differences in order, but they are essentially to avoid Null values caused by warp operations. Your observation is very careful. The difference is that you can re-extract features from the warped image, or directly warp features. For the mask, we prefer that it as a regional choice, so we do not extract mask from new patch again.
PS. _utils.transform()_ includes both warp and crop operations. @dkoguciuk

**Comment by JirongZhang** [comment]
For the feature, we first extract the features of the patch. After getting h, we warp the original image and crop it to get the new patch, and then extract the feature from the new patch.

For the mask, we first make mask prediction on the original image, and then cut the prediction results to fit the patch size. After H is obtained, we warp the whole mask and then cut it to fit the tensor size.

There are some differences in order, but they are essentially to avoid Null values caused by warp operations. Your observation is very careful. The difference is that you can re-extract features from the warped image, or directly warp features. For the mask, we prefer that it as a regional choice, so we do not extract mask from new patch again.

PS. utils.transform() includes both warp and crop operations. @dkoguciuk

**Comment by JirongZhang** [comment]
For the feature, we first extract the features of the patch. After getting homography, we warp the original image and crop it to get the new patch, and then extract the feature from the new patch.

For the mask, we first make mask prediction on the original image, and then cut the prediction results to fit the patch size. After homography is obtained, we warp the whole mask and then cut it to fit the tensor size.

There are some differences in order, but they are essentially to avoid Null values caused by warp operations. Your observation is very careful. The difference is that you can re-extract features from the warped image, or directly warp features. For the mask, we prefer that it as a regional choice, so we do not extract mask from new patch again.

PS. utils.transform() includes both warp and crop operations. @dkoguciuk

**Comment by dkoguciuk** [comment]
Hi @JirongZhang ,

I didn't understand your intuitions, to be honest. Did you try extracting mask from the transformed patch and your current approach was better? Or it's just a decision based on the intuitions?

Best,
Daniel

**Comment by JirongZhang** [comment]
It is a decision just based on intuitions. @dkoguciuk

**Comment by dkoguciuk** [comment]
@JirongZhang  - sure, thanks!
